### PR TITLE
fix: track established bonds

### LINKED ISSUE #4221: Support Enr responses even if the bucket is full in discv4
### Describe the feature

if the bucket is full then we don't add the peer to the bucket and are not able to handle enr requests:

https://github.com/paradigmxyz/reth/blob/7cd7859a2be3504ffcd61bdc385abee78ac26114/crates/net/discv4/src/lib.rs#L1147-L1156

instead we should keep track of all established bonds separately

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattsse**
Closes #4221

previously the discv4 impl was ignoring requests if the peer's bucket is full.
this now tracks established bonds separately and verifies bonds rather than bucket inclusion

```
./build/bin/devp2p --verbosity 5 discv4 requestenr enode://4aeb4ab6c14b23e2c4cfdce879c04b0748a20d8e9b59e25ded2a08143e265c6c25936e74cbc8e641e3312ca288673d91f2f93f8e277de3cfa444ecdaaf982052@157.90.35.166:30303
INFO [08-16|15:30:15.348] New local node record                    seq=1,692,192,615,348 id=224f5d8fef49a4b4 ip=127.0.0.1 udp=64908 tcp=0
TRACE[08-16|15:30:15.349] >> PING/v4                               id=6b36f791352f15eb addr=157.90.35.166:30303 err=nil
TRACE[08-16|15:30:15.381] << PONG/v4                               id=6b36f791352f15eb addr=157.90.35.166:30303 err=nil
TRACE[08-16|15:30:15.382] << PING/v4                               id=6b36f791352f15eb addr=157.90.35.166:30303 err=nil
TRACE[08-16|15:30:15.382] >> PONG/v4                               id=6b36f791352f15eb addr=157.90.35.166:30303 err=nil
TRACE[08-16|15:30:15.881] >> ENRREQUEST/v4                         id=6b36f791352f15eb addr=157.90.35.166:30303 err=nil
TRACE[08-16|15:30:15.906] << ENRRESPONSE/v4                        id=6b36f791352f15eb addr=157.90.35.166:30303 err=nil
enr:-Km4QLNlM6ax1LAVyEpfVB49kIczTWMgo7QPP7eWsuScLAGlY4bK33veSCLzT4234j2hu15ZM6ossxE_4rBxmwSn9i6GAX8OcbRig2V0aMfGhNzpbC2AgmlkgnY0gmlwhJ1aI6aDbGVzwQGJc2VjcDI1NmsxoQJK60q2wUsj4sTP3Oh5wEsHSKINjptZ4l3tKggUPiZcbIRzbmFwwIN0Y3CCdl-DdWRwgnZf
```

hive tests still pass:

image

### CODE REVIEW

File: crates/net/discv4/src/config.rs [hunk 1]
```diff
@@ -121,16 +123,17 @@ impl Default for Discv4Config {
             /// Every outgoing request will eventually lead to an incoming response
             udp_ingress_message_buffer: 1024,
             max_find_node_failures: 5,
-            ping_interval: Duration::from_secs(300),
+            ping_interval: Duration::from_secs(60 * 10),
```
Comment (gakonst) [inline crates/net/discv4/src/config.rs:126]: why'd this double? any impact

File: crates/net/discv4/src/lib.rs [hunk 2]
```diff
@@ -1543,10 +1562,15 @@ impl Discv4Service {
...
             // evict expired nodes
-            if self.evict_expired_requests_interval.poll_tick(cx).is_ready() {
+            while self.evict_expired_requests_interval.poll_tick(cx).is_ready() {
                 self.evict_expired_requests(Instant::now())
             }
 
+            // evict expired nodes
+            while self.expire_interval.poll_tick(cx).is_ready() {
+                self.received_pongs.evict_expired(Instant::now(), EXPIRE_DURATION)
+            }
```
Comment (gakonst) [inline crates/net/discv4/src/lib.rs:1572]: Why do we go from `if` to `while` here?

File: crates/net/discv4/src/lib.rs [hunk 3]
```diff
@@ -1144,8 +1164,7 @@ impl Discv4Service {
             return
         }
 
-        let key = kad_key(id);
-        if let BucketEntry::Present(_, _) = self.kbuckets.entry(&key) {
+        if self.has_bond(id, remote_addr.ip()) {
```
Comment (gakonst) [inline crates/net/discv4/src/lib.rs:1167]: Should this still include the old condition? Or just do that if we have a bond, and it's not possible to not have a bond and be present in the bucket?

File: crates/net/discv4/src/lib.rs [hunk 4]
```diff
@@ -928,13 +951,14 @@ impl Discv4Service {
...
-                        self.notify(DiscoveryUpdate::DiscoveredAtCapacity(record))
+                        self.notify(DiscoveryUpdate::DiscoveredAtCapacity(record));
+                        needs_bond = true;
                     }
-                    BucketInsertResult::FailedFilter |
-                    BucketInsertResult::TooManyIncoming |
-                    BucketInsertResult::NodeExists => {
+                    BucketInsertResult::TooManyIncoming | BucketInsertResult::NodeExists => {
+                        needs_bond = true;
                         // insert unsuccessful but we still want to send the pong
```
Comment (gakonst) [inline crates/net/discv4/src/lib.rs:959]: makes sense

File: crates/net/discv4/src/lib.rs [hunk 5]
```diff
@@ -1079,25 +1111,13 @@ impl Discv4Service {
...
-                }
-            }
-            kbucket::Entry::Absent(_) => {
-                // no existing endpoint proof
-                // > To guard against traffic amplification attacks, Neighbors replies should only be sent if the sender of FindNode has been verified by the endpoint proof procedure.
-            }
-            kbucket::Entry::SelfEntry => {}
+        if self.has_bond(node_id, remote_addr.ip()) {
+            self.respond_closest(msg.id, remote_addr)
         }
```
Comment (gakonst) [inline crates/net/discv4/src/lib.rs:1121]: No need to respond if `status.is_connected()` now? I guess it's implicit if you have a bond with the peer?

**Review by gakonst — APPROVED** [review]
Hopefully this fixes the Ethernodes / crawlers

File: crates/net/discv4/src/config.rs (hunk 1 above)
Comment (mattsse) [reply crates/net/discv4/src/config.rs:126]: the enforced bond duration is an hour, this should be sufficient to cycle through all of them in an hour

File: crates/net/discv4/src/lib.rs (hunk 2 above)
Comment (mattsse) [reply crates/net/discv4/src/lib.rs:1572]: poll_tick only registers on pending

File: crates/net/discv4/src/lib.rs (hunk 3 above)
Comment (mattsse) [reply crates/net/discv4/src/lib.rs:1167]: enforcing bond here is sufficient

**Pull request merged**